extensions/windows-event-log/CollectorInitiatedSubscription.cpp

```diff
@@ -0,0 +1,789 @@
+  }
+
+  if (vProperty->Type == EcVarTypeNull) {
+    LOG_SUBSCRIPTION_ERROR("!hArray");
+    return false;
+  }
+
+  EC_OBJECT_ARRAY_PROPERTY_HANDLE hArray = vProperty->PropertyHandleVal;
+  utils::ScopeGuard guard_hArray([hArray]() { EcClose(hArray); });
+
+  // Get the EventSources array size (number of elements).
+  DWORD dwEventSourceCount{};
+  if (!EcGetObjectArraySize(hArray, &dwEventSourceCount)) {
+    LOG_SUBSCRIPTION_WINDOWS_ERROR("EcGetObjectArraySize");
+    return false;
+  }
+
+  auto getArrayProperty = [this](EC_OBJECT_ARRAY_PROPERTY_HANDLE hArray, EC_SUBSCRIPTION_PROPERTY_ID propID, DWORD arrayIndex, DWORD flags, std::vector<BYTE>& buffer, PEC_VARIANT& vProperty)
+  {
+    buffer.clear();
+    buffer.resize(sizeof(EC_VARIANT));
+    DWORD dwBufferSizeUsed{};
+    if (!EcGetObjectArrayProperty(hArray, propID, arrayIndex, flags, (DWORD)buffer.size(), (PEC_VARIANT)&buffer[0], &dwBufferSizeUsed)) {
+      if (ERROR_INSUFFICIENT_BUFFER == GetLastError()) {
+        buffer.resize(dwBufferSizeUsed);
+        if (!EcGetObjectArrayProperty(hArray, propID, arrayIndex, flags, (DWORD)buffer.size(), (PEC_VARIANT)&buffer[0], &dwBufferSizeUsed)) {
+          LOG_SUBSCRIPTION_WINDOWS_ERROR("EcGetObjectArrayProperty");
+          return false;
+        }
+      }
+      else {
+        LOG_SUBSCRIPTION_WINDOWS_ERROR("EcGetObjectArrayProperty");
+        return false;
+      }
+    }
+
+    vProperty = (PEC_VARIANT)&buffer[0];
+
+    return true;
+  };
+
+  auto getStatus = [this](const std::wstring& eventSource, EC_SUBSCRIPTION_RUNTIME_STATUS_INFO_ID statusInfoID, DWORD flags, std::vector<BYTE>& buffer, PEC_VARIANT& vStatus) {
+    buffer.clear();
+    buffer.resize(sizeof(EC_VARIANT));
+    DWORD dwBufferSize{};
+    if (!EcGetSubscriptionRunTimeStatus(
+          s_subscriptionName.value().c_str(), 
+          statusInfoID, 
+          eventSource.c_str(),
+          flags,
+          (DWORD)buffer.size(),
+          (PEC_VARIANT)&buffer[0],
+          &dwBufferSize)) {
+      if (ERROR_INSUFFICIENT_BUFFER == GetLastError()) {
+        buffer.resize(dwBufferSize);
+        if (!EcGetSubscriptionRunTimeStatus(s_subscriptionName.value().c_str(),
+              statusInfoID,
+              eventSource.c_str(),
+              flags,
+              (DWORD)buffer.size(),
+              (PEC_VARIANT)&buffer[0],
+              &dwBufferSize)) {
+          LOG_SUBSCRIPTION_WINDOWS_ERROR("EcGetSubscriptionRunTimeStatus");
+        }
+      } else {
+        LOG_SUBSCRIPTION_WINDOWS_ERROR("EcGetSubscriptionRunTimeStatus");
+      }
+    }
+
+    vStatus = (PEC_VARIANT)&buffer[0];
+
+    return true;
+  };
+
+  for (DWORD i = 0; i < dwEventSourceCount; i++) {
+    std::vector<BYTE> eventSourceBuffer;
+    PEC_VARIANT vProperty = NULL;
+    if (!getArrayProperty(hArray, EcSubscriptionEventSourceAddress, i, 0, eventSourceBuffer, vProperty)) {
+      return false;
+    }
+
+    if (vProperty->Type != EcVarTypeNull && vProperty->Type != EcVarTypeString) {
```

Comment:
Are we sure vProperty will not be NULL here?

Reply:
If `getArrayProperty` returns true, `vProperty` points to first element of `eventSourceBuffer`.